### PR TITLE
1821 Colorize the mutation points bar labels 

### LINKED ISSUE #1821: Colorize the mutation points bar labels
They are currently colored white and a bit hard to read against the full mutation points bar. In the concepts the texts are supposed to look like this:
9961f0310c40b3a504e8ef8f4170ca7854aa7f5c
This is a bit difficult because that color made them almost invisible against the bar background when the it's empty (the bar has black bg ingame). I think it would be nice if the labels could automatically adjust their colors according to the current background they are on if that's possible in godot (maybe with shaders?).

### PR DESCRIPTION
**Pull request by FuJa0815**
grafik
This effect is achieved with a shader.
Closes #1821

### CODE REVIEW

**Review by Kasterisk — COMMENTED** [review]
Suggestion: https://github.com/Revolutionary-Games/Thrive/issues/1821#issuecomment-728126981

File: shaders/BooleanInverseColorShader.shader [hunk 1]
```diff
@@ -0,0 +1,13 @@
+shader_type canvas_item;
+
+uniform vec4 fg_col : hint_color = vec4(0.745098, 0.356863, 0.921569, 1);
+uniform vec4 bg_col : hint_color = vec4(0.290196, 0.14902, 0.360784, 1);
```
Comment (Kasterisk) [inline shaders/BooleanInverseColorShader.shader:4]: Add a blank line

**Review by Kasterisk — CHANGES_REQUESTED** [review]
A minor styling issue, other than that I think this works nicely.

File: shaders/BooleanInverseColorShader.shader [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+void fragment() {
+    vec4 screen_col = textureLod(SCREEN_TEXTURE, SCREEN_UV, 0.0);
+    float alpha = texture(TEXTURE, UV).a;
+    if ((fg_col - screen_col).x < 0.1) {
+        COLOR = vec4(bg_col.xyz, alpha);
+    } else {
+        COLOR = vec4(fg_col.xyz, alpha);
+    }
+}
```
Comment (hhyyrylainen) [inline shaders/BooleanInverseColorShader.shader:13]: Add a blank line here.

I think this file should be renamed, as right now this seems to not in fact inverse the colour, instead this selects one of two colours based on how far they are from what the current colour is.
Also has this been tested with gles2 mode? A basic GUI part should be tested to work fine there as well.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: shaders/BooleanInverseColorShader.shader (hunk 2 above)
Comment (Kasterisk) [reply shaders/BooleanInverseColorShader.shader:13]: Gles2 mode works fine with this in my laptop

Comment (hhyyrylainen) [reply shaders/BooleanInverseColorShader.shader:13]: That's good, I remember there being some post somewhere, that (maybe it was older version) Godot gles2 shaders were not allowed to read the screen, or maybe it was that it causes just a performance impact now?

Comment (hhyyrylainen) [reply shaders/BooleanInverseColorShader.shader:13]: If there is a major performance impact (more than like 5 FPS at around 60 FPS), I think we need a performance option to not use the shader.

Comment (Kasterisk) [reply shaders/BooleanInverseColorShader.shader:13]: I don't see any noticeable decrease in fps tbh (at least for me), maybe someone else could also test it out

Comment (hhyyrylainen) [reply shaders/BooleanInverseColorShader.shader:13]: I think that's good enough, I may be just remembering some old posts that screen reading shaders were really bad in GLES2, we can add an option later if needed to disable the effect.

**Comment by Zireael07** [comment]
Wondering why you felt the need to use a shader here instead of checking for the value (how much of the bar is filled)

**Comment by hhyyrylainen** [comment]
It will be useful in general to have a component for making a Control pick a suitable colour automatically. Also due to translations, GUI scaling to different window resolutions, it would be flaky solution at best to explicitly make the colour change based on the value of the bar.

**Comment by FuJa0815** [comment]
@hhyyrylainen have you any better idea regarding the shader name?
I mean, it doesn't inverse the color itself, but it "nots" the color. It checks the existing color and selects the other color.

**Comment by hhyyrylainen** [comment]
I think something like "SelectDistinctFromBackgroundColour" could be a suitable name for what it does.

**Comment by Kasterisk** [comment]
The shader file is missing, maybe you forgot to commit it?

**Review by Kasterisk — APPROVED** [review]
Looks good!

File: src/microbe_stage/editor/MutationPointsText.tres [hunk 3]
```diff
@@ -0,0 +1,9 @@
+[gd_resource type="ShaderMaterial" load_steps=2 format=2]
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MutationPointsText.tres:1]: I think this file is a material? If so this should be named "MutationPointsTextMaterial.tres"

File: shaders/SelectDistinctFromBackgroundColor.shader [hunk 4]
```diff
@@ -0,0 +1,17 @@
+shader_type canvas_item;
+
+uniform vec4 bg_col : hint_color;
```
Comment (hhyyrylainen) [inline shaders/SelectDistinctFromBackgroundColor.shader:3]: please don't abbreviate "colour"

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Two small comments. Otherwise this seems fine.

File: shaders/SelectDistinctFromBackgroundColor.shader (hunk 4 above)
Comment (FuJa0815) [reply shaders/SelectDistinctFromBackgroundColor.shader:3]: speaking of colo(u)r: are there any guidelines on british vs american english?

Comment (hhyyrylainen) [reply shaders/SelectDistinctFromBackgroundColor.shader:3]: Use British English, unless something like Godot requires to use American spelling.

File: shaders/SelectDistinctFromBackgroundColor.shader [hunk 5]
```diff
@@ -0,0 +1,17 @@
+shader_type canvas_item;
+
+uniform vec4 bg_colour : hint_color;
+
+uniform vec4 main_colour : hint_color;
+uniform vec4 sec_colour : hint_color;
```
Comment (hhyyrylainen) [inline shaders/SelectDistinctFromBackgroundColor.shader:6]:
```suggestion
uniform vec4 backgroundColour : hint_color;

uniform vec4 mainColour : hint_color;
uniform vec4 secondaryColour : hint_color;
```
for consistent naming with other shaders

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: shaders/SelectDistinctFromBackgroundColor.shader (hunk 5 above)
Comment (hhyyrylainen) [reply shaders/SelectDistinctFromBackgroundColor.shader:6]: You didn't commit a fix for this?

Comment (FuJa0815) [reply shaders/SelectDistinctFromBackgroundColor.shader:6]: Whops, didn't wait for the console output of git warning me about unpulled changes :D

**Review by hhyyrylainen — APPROVED** [review]

**Pull request merged**